src/twisted/internet/defer.py

```diff
@@ -159,6 +165,65 @@ def maybeDeferred(f, *args, **kw):
         return succeed(result)
 
 
+def unbox(deferred, raises=True, passthrough=True):
```

Comment:
The reason I added ``raise=False`` was when I went to implement ``successResultOf``, ``failureResultOf``, and ``assertNoResult`` in terms it was hard to replicate the behavior of those functions without it:

* ``successResultOf`` uses the failure's traceback printing mechanisms which seem to be subtly different than Python's built in ones. If we're happy with the output on failure changing slightly then this one isn't a problem.
* ``failureResultOf`` returns the failure, which is impossible to do if ``unbox()`` doesn't provide that information to it.
* ``failureResultOf`` when there are expected exception types has the same problem as ``sucessResultOf`` with the failure message relying on details of failure's traceback formatting.
* ``assertNoResult`` wants to silence failures but *not* successes, so the current ``passthrough`` parameter doesn't help it here, but a ``consumeErrors`` would.

In my own personal use cases I'm hoping to use the ``unbox()`` function in functions that can either be driven synchronously or asynchronously depending on what "fetcher" is passed into them (either sync or async). I want to just always return ``Deferred`` from these functions but I need a way to hide the fact that ``Deferreds`` are involved at all for end users who want to call the code synchronously. The expected way for this to look is something like (simplified):

```python
import requests
import treq

def sync_fetcher(url):
    resp = requests.get(url)
    resp.raise_for_status()
    return resp.content

def async_fetcher(url):
    d = treq.get(url)
    d.addCallback(treq.content)
    return d


def api(thing, fetcher):
    d = Deferred()
    d.addCallback(fetcher)
    d.callback("https://example.com/api/{}".format(thing))
    return d


# When being used from async code
d = api("my-thing", fetcher=async_fetcher)
d.addCallback(print)

# When being used from sync code
print(unbox(api("my-thing", fetcher=sync_fetcher)))

```

The goal is to essentially treat the ``unbox()`` function sort of like string encoding/decoding in that it turns the code into "normal" sync like code that hides the fact ``Deferreds`` are in play. However, if errors don't get consumed then the above code is going to print a spurious error to stdout when the ``Deferred`` gets garbage collected. The way to work around this would be:

```python
d = api("my-thing", fetcher=sync_fetcher)
try:
    print(unbox(d))
finally:
    d.addErrBack(lambda _: None)
```

Which fails at my goal of being able to tell people with "normal" sync looking code to just wrap the ``Deferred`` returning functions with an ``unbox()`` and forget about it.